### PR TITLE
BREAKING: add firefox support

### LINKED ISSUE #24: Is there any way to run Dawn without Chrome?
## Summary

It seems that Dawn needs Chrome to work. However, even with chrome, it doesn't seem to work. 

## Steps To Reproduce The Bug
I have chrome installed (on Windows), and I try to run the example ts file in the README, app_test.ts:
```sh
deno test --allow-run app_test.ts
```
I get failures for both, even though one should pass (ignore my horribly misconfigured prompt):
app_test_fail

According to the first test, it should pass, but it doesn't. If we look at the stack trace, we find three relevant files: app_test.ts, dawn.ts, and headless_browser.ts. Since headless_browser.ts is the relevant file closest to the top of the stack, it is most likely that the error is there. If we take a look at headless_browser.ts:
https://github.com/drashland/dawn/blob/833b6a1d118868289d03129bb9e3cf898f75dd74/src/headless_browser.ts#L101-L108

So it appears that Deno is executing "start chrome", but for some reason, it can't find it? However, I have chrome, and 'start chrome' seems to work when I run it from the command prompt: 
start_chrome

## Expected Behavior
I expect dawn to run app_test.ts with the first test passing and the second test failing. 

## Suggested Solution(s)
My best hypothesis is that Deno doesn't inherit the PATH from Command Prompt. I still think it would be really cool to let Dawn run without requiring Chrome.

### PR DESCRIPTION
**Pull request by ebebbington**
Fixes #24 

## Summary

* To support firefox, the API has been changed slightly, this makes both chrome and firefox classes maintainable. Tests have been adjusted for this and it was pretty quick and easy to do luckily
* API's for firefox and chrome client are exactly the same
* Requires a new permission: `--allow-write` if using firefox. example: `deno run --allow-read=$TMPDIR app_test.ts`
* Options can be passed to the `build` method, to customise the client
* The API is now as follows:
```ts
import { buildFor } from  "https://deno.land/x/sinco/mod.ts"

const Sinco = await buildFor("chrome", ...opts) // or replace "chrome" with "firefox"
await Sinco.goTo("https://drash.land") // as usual
```
* Added firefox support
* Removed waitForAnchorChange method

### CODE REVIEW

**Comment by ebebbington** [comment]
so im debating removing the `waitForAnchorChange`, because if i go to chromestatus.com, and change the value of the input field we modify in our tests, the uri doesn't change, but the chrome client does pick it up

Alongside this, if i do the same with a firefox client, the client doesn't pick up any event for this

File: tests/integration/docker_test/drivers.dockerfile [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+
+#RUN apt --no-install-recommends -y install firefox
+#RUN rm -rf /var/lib/apt/lists/* /var/cache/apt/* \
+#  && wget --no-verbose -O /tmp/firefox.tar.bz2 https://download-installer.cdn.mozilla.net/pub/firefox/releases/$FIREFOX_VERSION/linux-x86_64/en-US/firefox-$FIREFOX_VERSION.tar.bz2 \
+#  && apt -y purge firefox \
+#  && rm -rf /opt/firefox \
+#  && tar -C /opt -xjf /tmp/firefox.tar.bz2 \
+#  && rm /tmp/firefox.tar.bz2 \
+#  && mv /opt/firefox /opt/firefox-$FIREFOX_VERSION \
+#  && ln -fs /opt/firefox-$FIREFOX_VERSION/firefox /usr/bin/firefox
```
Comment (saragee3) [inline tests/integration/docker_test/drivers.dockerfile:62]: what is all this?

File: src/firefox_client.ts [hunk 2]
```diff
@@ -0,0 +1,658 @@
...
+    return packet;
+  }
+
+  /**
+   * Get full path to the firefox binary on the user'ss filesystem.
+   * Thanks to [caspervonb](https://github.com/caspervonb/deno-web/blob/master/browser.ts)
+   *
+   * @returns the path
+   */
+  private static getFirefoxPath(): string {
```
Comment (saragee3) [inline src/firefox_client.ts:648]: what if the machine doesn't have FF installed?

File: src/firefox_client.ts [hunk 3]
```diff
@@ -0,0 +1,658 @@
...
+      try {
+        const j = JSON.parse(partial);
+        partial = "";
+        packets.unshift(j);
+      } catch (err) {
+        // still a partial, do nothing
+      }
+
+      const validPackets = packets.filter((packet) => {
+        if (UNSOLICITED_EVENTS.includes(packet.type) === true) {
```
Comment (saragee3) [inline src/firefox_client.ts:504]: so `UNSOLICITED_EVENTS` are silently ignored? maybe we can provide an FF doc link on line 3 so the events are easily accessible

File: tests/integration/docker_test/drivers.dockerfile (hunk 1 above)
Comment (ebebbington) [reply tests/integration/docker_test/drivers.dockerfile:62]: All the possible ways to install a driver that never fully worked lol, tbh I only kept them for reference but I think it’s a similar situation to that Drash 1000 resource script thingy - I’ll remove it then :)

File: src/firefox_client.ts (hunk 2 above)
Comment (ebebbington) [reply src/firefox_client.ts:648]: It’s going to throw a file not found error when it tries to run Firefox from the sub process

File: src/firefox_client.ts (hunk 3 above)
Comment (ebebbington) [reply src/firefox_client.ts:504]: If only I knew what the events were xD that unsolicited events variable was actually pulled from a node Firefox package, and I just added to it when I found an event we don’t care about, but I’ll try dig around for some docs

File: tests/integration/docker_test/drivers.dockerfile (hunk 1 above)
Comment (ebebbington) [reply tests/integration/docker_test/drivers.dockerfile]: https://github.com/drashland/sinco/pull/52/commits/9a5ffe31407111f3d41c35e63c1f719dcb255736 (decided to save it under my GH for future ref instead)

File: src/firefox_client.ts (hunk 3 above)
Comment (crookse) [reply src/firefox_client.ts:506]: hmm... since it came from a node package, maybe we just state what node package it came from and that they're ignored

Comment (ebebbington) [reply src/firefox_client.ts:506]: Yeah alright sounds fair

Comment (ebebbington) [reply src/firefox_client.ts:506]: I can't even find their source code for it now, maybe they removed it but im going to add a note anyways

Comment (crookse) [reply src/firefox_client.ts:506]: okie doke

Comment (ebebbington) [reply src/firefox_client.ts:513]: https://github.com/drashland/sinco/pull/52/commits/020ae1577d5f44242069c780adfe4a2a303ce999

**Review by crookse — APPROVED** [review]

**Pull request merged**